[Merged by Bors] - feat(combinatorics): define simple graphs

awainverse opened the pull request:
adds basic definition of `simple_graph`s

---
<!-- put comments you want to keep out of the PR commit here -->

jalex-stark committed:
init

jalex-stark committed:
remove extra file

jalex-stark committed:
variables in sym_matrix

jalex-stark committed:
imports, friendship

awainverse committed:
Hopefully updating char_poly.lean enough

jalex-stark committed:
fixed broken proof

jalex-stark committed:
golf

jalex-stark committed:
remove unnecessary line

jalex-stark committed:
Merge branch 'freek_83_friendship_theorem' of https://github.com/leanprover-community/mathlib into freek_83_friendship_theorem

jalex-stark committed:
import golf

jalex-stark committed:
renaming

jalex-stark committed:
clean proofs

jalex-stark committed:
separate simple_graph

jalex-stark committed:
golf

jalex-stark committed:
rename

jalex-stark committed:
..and remove the old file

jalex-stark committed:
golf

jalex-stark committed:
golf

jalex-stark committed:
optimize main proof

jalex-stark committed:
unified `pow`s

jalex-stark committed:
golf

jalex-stark committed:
Merge remote-tracking branch 'origin/master' into freek_83_friendship_theorem

jalex-stark committed:
formatting

jalex-stark committed:
golf

jalex-stark committed:
golf

jalex-stark committed:
copyright

jalex-stark committed:
explicit arguments

jalex-stark committed:
formatting

awainverse committed:
Kyle's graph nomenclature

awainverse committed:
Edge set

jalex-stark committed:
rename

jalex-stark committed:
Merge branch 'freek_83_friendship_theorem' of https://github.com/leanprover-community/mathlib into freek_83_friendship_theorem

jalex-stark committed:
move into mathlib

awainverse committed:
Just accepting the docstrings?

jalex-stark committed:
rename A and B to left_ and right_carrier

jalex-stark committed:
implementation details

awainverse committed:
Renaming

awainverse committed:
Merge branch 'freek_83_friendship_theorem' of https://github.com/leanprover-community/mathlib into freek_83_friendship_theorem

awainverse committed:
Finish that renaming

jalex-stark committed:
renaming

jalex-stark committed:
Merge branch 'freek_83_friendship_theorem' of https://github.com/leanprover-community/mathlib into freek_83_friendship_theorem

jalex-stark committed:
add comment

jalex-stark committed:
contract

jalex-stark committed:
compress

awainverse committed:
left and right

awainverse committed:
Merge branch 'freek_83_friendship_theorem' of https://github.com/leanprover-community/mathlib into freek_83_friendship_theorem

jalex-stark committed:
Merge branch 'freek_83_friendship_theorem' of https://github.com/leanprover-community/mathlib into freek_83_friendship_theorem

jalex-stark committed:
Merge branch 'freek_83_friendship_theorem' into graphs_and_bigraphs

jalex-stark committed:
remove archive stuff

bryangingechen commented on src/combinatorics/bigraph.lean:
```diff
@@ -0,0 +1,289 @@
+import algebra.big_operators
+
+/-!
+# Bigraphs
```
The term "bipartite graph" should appear somewhere here. Is bigraph a standard abbreviation? The [wikipedia page](https://en.wikipedia.org/wiki/Bigraph) is about something else.

awainverse replied:
It pops up, for instance: https://mathworld.wolfram.com/Bigraph.html. I had not heard of the concept on that wikipedia page.

I got the term from page 27 of this model-theory paper https://www.math.ucla.edu/~matthias/pdf/VC%20Density,%20I.pdf
They do cite combinatorialists for that distinction, although I haven't checked out the cited paper yet.

The distinction is basically that a `bigraph` has a specified left part and right part, while a bipartite graph just can be partitioned... in my mind, a bipartite graph is just a graph with no odd cycles.

bryangingechen replied:
Ah, I see. All this is good info to include in the module doc!

jalex-stark replied:
I put a little discussion in the module docstring, without renaming the file or the declarations. I'm happy to call everything `bipartite_graph` instead if that is preferred.

bryangingechen replied:
I'm fine with these being called `bigraph`s . I do think the part about there being a specified left and right part is important, so it should also be added to the module docs (and the docstrings when you get to them).

If you can add some explicit citations for this concept / name that'd be great too.

jalex-stark committed:
clarify naming choice

jalex-stark committed:
lint

jalex-stark committed:
update docs

jalex-stark committed:
dedup paragraph

robertylewis commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,80 @@
+
+end finite
+end simple_graph
+#lint
```
```suggestion
```

robertylewis committed:
Update src/combinatorics/simple_graph.lean

awainverse committed:
Only simple graphs for now

awainverse committed:
Merge branch 'graphs_and_bigraphs' of https://github.com/leanprover-community/mathlib into graphs_and_bigraphs

kmill committed:
Various modifications to simple_graphs PR

Made things decidable (removed nonclassical)

Changed how degree works.  Made it use `fintype` rather than
`finset`. This seems more general.

Made `regular` depend only on local finiteness, generalized to
infinite graphs.

Define complete graphs, used it for inhabited instance.

Proved the complete graph is regular.

Defined edge incidence

kmill committed:
Switched neighbors back to finset interface and renamed it to adj'

awainverse committed:
Renamed back to neighbors, modified the sections

kmill committed:
Some docstrings and decidable instance manipulation

kmill committed:
Forgot to commit a few more source beautifications

awainverse committed:
Integrating Kyle's contributions

jalex-stark committed:
naming

jalex-stark committed:
Merge remote-tracking branch 'origin/graphs_and_bigraphs_2' into graphs_and_bigraphs

jalex-stark committed:
naming and golfing

jalex-stark committed:
just a little more golf

awainverse committed:
has_mem nomenclature

awainverse committed:
docstring

jalex-stark committed:
spacing, golfing

jalex-stark committed:
Merge branch 'graphs_and_bigraphs' of https://github.com/leanprover-community/mathlib into graphs_and_bigraphs

jalex-stark committed:
...and more spacing

jalex-stark committed:
decidability golf

kmill reviewed (commented):
It looks good to me :+1:  (No comment on mathlib style or naming -- that's for the experts)

jalex-stark committed:
more golf

jalex-stark committed:
Merge remote-tracking branch 'origin/master' into graphs_and_bigraphs

robertylewis commented on src/data/sym2.lean:
```diff
@@ -282,4 +292,22 @@ equiv_sym α
+/--
+Given `[decidable_eq α]` and `[fintype α]`, the following instance gives `fintype (sym2 α)`.
+-/
+instance (α : Type*) [decidable_eq α] : decidable_rel (sym2.rel α) :=
```
I'm slightly worried by this. `decidable` is data and this could potentially be used to compute with. We try to avoid defining data in tactic mode because it's not always clear what you're getting. 

The "big guns" here (`cc`) look like they're only used in the `Prop` part, which is fine. But I'd be more comfortable if this were slightly refactored so the algorithmic part was clearer and avoided `rintro` and `subst`.

awainverse replied:
I may have to either leave this for @kmill or just make everything noncomputable for now and wait for Kyle to retrofit it in a separate PR

kim-em commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,176 @@
+a `has_mem` instance.
+-/
+structure simple_graph (V : Type u) :=
+(adj : V → set V)
```
I think I'm unconvinced by this. Why do we want the mem notation at all? I think it's best if the API commits to one or the other (i.e. always write `G.adj v w`, in which case this should be `V -> V -> Prop`, or always write `w \mem G.adj v`). Otherwise it's inviting hassle as people need to switch syntax.

kim-em commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,176 @@
+The complete graph on a type `V` is the simple graph with all pairs of distinct vertices adjacent.
+-/
+def complete_graph (V : Type u) : simple_graph V :=
+{ adj := λ v w, v ≠ w }
```
Yeah, see even here you're abusing the set API.

(My suggestion boils down to: ideally, `set V := V -> Prop` would be an irreducible definition!)

kim-em commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,176 @@
+The edges of G consist of the unordered pairs of vertices related by
+`G.adj`.  It is given as a subtype of the symmetric square.
+-/
+def E : Type u := {x : sym2 V // x ∈ sym2.from_rel G.sym}
```
Again, I'm not super keen on this, because it requires everyone to "know the definition" to work with it (i.e. to know that you're meant to use it as a subtype).

Could we add a `has_mem` instance, provide a few iff lemmas, then mark is as irreducible before continuing further?

kim-em replied:
(i.e. so the "implementation detail" involving `subtype` and `sym2` is completely hidden from the rest of the file, and hence from any users of the graph API?)

kim-em replied:
(I'll admit to being picky here --- I have a new found enthusiasm for `@[irreducible]`, but it's okay to shout me down.)

jalex-stark replied:
We did something here. It's possible that we still need another lemma or two before the `irreducible` attribute is declared, but we won't know what they are until we use the current API more.

kmill replied:
I was away this week on vacation, so I'm now getting around to reviewing the changes.  Everything's seemed good, but this one I'm not sure about.  My preference is for `E` to have the definition
```
def E : set (sym2 V) := sym2.from_rel G.sym
```
and I'm not sure exactly at which point it became a subtype (it may or may not have been me). This is essentially the classical definition -- the edge set is a subset of unordered pairs on the vertex set.

If we make `E` irreducible, then we'd eventually be copying the whole interface from `data.sym2` (most of the interface for edges, except for `adj_iff_exists_edge`, are thin wrappers), and we'd have been better off defining edges from scratch in this file.  A reason not to do that is that, for example, multigraphs are likely going to use `sym2` as well in some capacity.

kim-em commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,176 @@
+
+/--
+`G.incident_edges v` is the set of edges incident to `v`.  Due to the way
+sets are defined, `G.incident_edges v e` denotes that `v` is incident to `e`.
```
Please don't encourage people to break the set API. :-)

awainverse replied:
I think it's being respected now...

awainverse committed:
Trying to abuse set less

jalex-stark committed:
new lemma

jalex-stark committed:
Merge branch 'graphs_and_bigraphs' of https://github.com/leanprover-community/mathlib into graphs_and_bigraphs

jalex-stark committed:
proved lemma

jalex-stark committed:
add attribute irreducible to G.E

awainverse committed:
Name change

awainverse committed:
Simplify complete graph definition

jalex-stark committed:
E.other

jalex-stark committed:
Merge branch 'graphs_and_bigraphs' of https://github.com/leanprover-community/mathlib into graphs_and_bigraphs

jalex-stark committed:
fix up decidable instance

jalex-stark committed:
undo bad simp lemma

jalex-stark committed:
add missing lemma

jalex-stark committed:
remove extra comment

awainverse committed:
a smidge of term mode

jalex-stark commented on src/data/sym2.lean:
```diff
@@ -167,7 +167,7 @@ begin
-lemma mem_iff {a b c : α} : a ∈ ⟦(b, c)⟧ ↔ a = b ∨ a = c :=
+@[simp] lemma mem_iff {a b c : α} : a ∈ ⟦(b, c)⟧ ↔ a = b ∨ a = c :=
 { mp  := by { rintro ⟨_, h⟩, rw eq_iff at h, tidy },
   mpr := by { rintro ⟨_⟩; subst a, { apply mk_has_mem }, rw eq_swap, apply mk_has_mem } }
```
```suggestion
  mpr := by { rintro ⟨_⟩; subst a, { apply mk_has_mem }, apply mk_has_mem_right } }
```

jalex-stark committed:
Update src/data/sym2.lean

bryangingechen commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,226 @@
+
+open finset
+-- noncomputable theory
+
```
```suggestion
```

bryangingechen commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,226 @@
+simple graph.
+
+-/
+
```
```suggestion
open finset
```

bryangingechen commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,226 @@
+Note: The type of the relation is given as `V → set V` rather than
+`V → V → Prop` so that, given vertices `v` and `w`, then `w ∈ G.adj v`
+works as another way to write `G.adj v w`.  Otherwise Lean cannot find
+a `has_mem` instance.
```
```suggestion
```

bryangingechen commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,226 @@
+/--
+A regular graph is a locally finite graph such that every vertex has the same degree.
+-/
+def regular_graph (d : ℕ) : Prop := ∀ (v : V), G.degree v = d
```
```suggestion
A locally finite simple graph is regular of degree `d` if every vertex has degree `d`.
-/
def is_regular_of_degree (d : ℕ) : Prop := ∀ (v : V), G.degree v = d
```

bryangingechen replied:
Not 100% sold on the new name suggestion, as it's a bit long. We do generally name `Prop`s `is_X` rather than just `X` though.

jalex-stark replied:
in maths we might say `is_d_regular`, but i feel like we can't use `d` in the name if `d` appears only as a bound variable rather than as something typeful.

If we made it a class, then `[d_regular G d]` would be similar to `[char_p R p]`

jalex-stark replied:
actually I find `G.is_regular_of_degree d` pretty aesthetically nice, and it's not going to appear inside of longer terms very much.

bryangingechen commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,226 @@
+end
+
+lemma complete_graph_is_regular [decidable_eq V] :
+  regular_graph (complete_graph V) (fintype.card V - 1) :=
```
```suggestion
  (complete_graph V).is_regular_of_degree (fintype.card V - 1) :=
```

bryangingechen commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,226 @@
+variables [fintype V]
+
+instance neighbor_set_fintype [decidable_rel G.adj] (v : V) : fintype (G.neighbor_set v) :=
+  @subtype.fintype _ _ (by {simp_rw mem_neighbor_set, apply_instance}) _
```
```suggestion
  @subtype.fintype _ _ (by { simp_rw mem_neighbor_set, apply_instance }) _
```

bryangingechen commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,226 @@
+  w ∈ G.edge_of_adj h := sym2.mk_has_mem_right v w
+
+lemma adj_iff_exists_edge {v w : V} (hne : v ≠ w) :
+G.adj v w ↔ ∃ (e : G.E), v ∈ e ∧ w ∈ e :=
```
```suggestion
  G.adj v w ↔ ∃ (e : G.E), v ∈ e ∧ w ∈ e :=
```

bryangingechen commented on src/data/sym2.lean:
```diff
@@ -282,4 +292,33 @@ equiv_sym α
+  rotate 2, { contrapose! h, cases h; cc },
+  all_goals { subst x₁, split; intro h1,
+    { subst h1; apply sym2.rel.swap },
+    { cases h1; cc }}
```
```suggestion
    { cases h1; cc } }
```

jalex-stark committed:
Update src/data/sym2.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

jalex-stark committed:
Update src/combinatorics/simple_graph.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

jalex-stark committed:
Update src/combinatorics/simple_graph.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

jalex-stark committed:
Update src/combinatorics/simple_graph.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

jalex-stark committed:
Update src/combinatorics/simple_graph.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

jalex-stark committed:
Update src/combinatorics/simple_graph.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

jalex-stark committed:
Update src/combinatorics/simple_graph.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

jalex-stark committed:
Update src/combinatorics/simple_graph.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

awainverse committed:
init

awainverse committed:
Sorry, meant to make a new branch

kim-em commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,217 @@
+Use `neighbor_finset_eq_filter` to rewrite this definition as a `filter`.
+-/
+
+variables (v : V) [fintype (G.neighbor_set v)]
```
Why not avoid the coercion and use `set.finite` here? It seems a bit unpleasant to define this as a `set` but then use `fintype`.

kmill replied:
This lets `G.degree v` work without needing to pass a finiteness proof.  For example, the instance `neighbor_set_fintype` a bit later on means `G.degree v` can find all the `fintype` instances it needs for finite graphs.

Also, isn't `set.fintype s` defined as `nonempty (fintype s)`?  One goal for the graph library is to be able to eventually prove basic graph algorithms are correct, and it seems like using `nonempty` might make things noncomputable.

kim-em commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,217 @@
+
+section locally_finite
+
+variable [∀ (v : V), fintype (G.neighbor_set v)]
```
How about defining `locally_finite G := ∀ (v : V), fintype (G.neighbor_set v)`?

(or whatever is appropriate if you change away from `fintype`)

kmill replied:
We had some trouble defining this before, but it seems we missed the correct attribute:
```
/--
A graph is locally finite if every vertex has a finite neighbor set.
-/
@[reducible]
def locally_finite := Π (v : V), fintype (G.neighbor_set v)

variable [locally_finite G]
```

If I remember correctly, the issue was that finiteness instances weren't being inferred without `reducible`.

kmill replied:
Ok, added this.

kim-em commented on src/data/sym2.lean:
```diff
@@ -282,4 +292,33 @@ equiv_sym α
 
 end sym_equiv
 
+section finite
```
```suggestion
section fintype
```

kim-em commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,217 @@
+  exact G.ne_of_edge key,
+end
+
+attribute [irreducible] E.other
```
Could you add a note saying why you do this?

kim-em commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,217 @@
+⟨begin
+  rcases inhabited.default {p : V × V | G.adj p.1 p.2} with ⟨⟨x, y⟩, h⟩,
+  use ⟦(x, y)⟧, rwa sym2.from_rel_prop,
+end⟩
```
What purpose does this serve?

kmill replied:
Without it, the linter says
```
/- TYPES ARE MISSING INHABITED INSTANCES: -/
#print simple_graph.E /- inhabited instance missing -/
```

kim-em replied:
I see. This is a bit of a pain. @fpvandoorn, do you have advice here on how to balance satisfying the `inhabited` linter, without cluttering up the library with a meaningless result? Would it be better to for example show that the edge set for the complete graph on two vertices was inhabited?

I used to hate the inhabited linter, but I realised that often enough it was prompting me to write "the first non-trivial example" and that this was often a useful exercise (typically because it showed you how terrible the API you'd just written was...).

kmill replied:
We just changed `simple_graph.E` to be a set `simple_graph.edge_set`.  The `E` type will reappear later when we create the multigraph interface.  But for now, means there's no need for the inhabited instance anymore, so it has been removed.

kim-em commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,217 @@
+
+/--
+The edges of G consist of the unordered pairs of vertices related by
+`G.adj`.  It is given as a subtype of the symmetric square.
```
```suggestion
The edges of G consist of the unordered pairs of vertices related by
`G.adj`.  

While it is defined here as a subtype of the symmetric square, after building an API for it below, we mark it [irreducible] so the particular representation is not important.
```

kmill committed:
Switch edge set to be a set

kmill committed:
Changed simple_graph.E to be a set simple_graph.edge_set

This means there is no need to duplicate the `sym2` interface within `simple_graph`.

Also incorporated some of Scott Morrison's suggestions.

Added `edge_finset` as well, since that can be useful for big_operators.

kmill committed:
Merge branch 'master' into graphs_and_bigraphs
merge master

jcommelin commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,202 @@
+/-- `G.neighbor_set v` is the set of vertices adjacent to `v` in `G`. -/
+def neighbor_set (v : V) : set V := set_of (G.adj v)
+
+lemma ne_of_edge {a b : V} (hab : G.adj a b) : a ≠ b :=
```
I think it would be good to try to standardize terminology: `edge` vs `adj` (I can think of a hypothetical dialect of English where they would be pronounced the same :rofl:).
In this case I would expect `ne_of_adj`. Should `adj` be renamed to `edge` completely? Or do you want to reserve `edge` for the unordered notion? In that case this should be `ne_of_adj` and `ne_of_edge` will be a different lemma.

kmill replied:
I would expect `ne_of_adj`, too.

It seems nice having `G.adj` be the adjacency relation and edges being elements of the edge set.

For simple graphs, maybe it is OK if `edge` and `adj` might be pronounced the same since they are similar concepts? :laughing:

awainverse replied:
This specific lemma is renamed, but I'm not renaming it as resolved because of the discussion about renaming `adj` to `edge`, which I feel neutral about.

jcommelin commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,202 @@
+lemma edge_iff_adj {v w : V} : ⟦(v, w)⟧ ∈ G.edge_set ↔ G.adj v w :=
+by refl
+
+lemma adj_iff_exists_edge {v w : V} (hne : v ≠ w) :
```
You could move `hne` to the right hand side of the `iff`, right?

kmill replied:
Yes, that's true, and it should probably be moved there.

jcommelin commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,202 @@
+variables [fintype V]
+
+instance neighbor_set_fintype [decidable_rel G.adj] (v : V) : fintype (G.neighbor_set v) :=
+  @subtype.fintype _ _ (by { simp_rw mem_neighbor_set, apply_instance }) _
```
```suggestion
@subtype.fintype _ _ (by { simp_rw mem_neighbor_set, apply_instance }) _
```

awainverse committed:
space

Co-authored-by: Johan Commelin <johan@commelin.net>

awainverse committed:
Reviews

awainverse committed:
Reviews

fpvandoorn commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,202 @@
+def edge_set : set (sym2 V) := sym2.from_rel G.sym
+
+@[simp]
+lemma edge_iff_adj {v w : V} : ⟦(v, w)⟧ ∈ G.edge_set ↔ G.adj v w :=
```
I think this should be renamed to `mem_edge_set` or `mem_edge_set_iff`.

kmill replied:
Changed it to `mem_edge_set`

kmill replied:
(To record why: it has parallelism with `mem_neighbor_set`.)

fpvandoorn commented on src/combinatorics/simple_graph.lean:
```diff
@@ -0,0 +1,202 @@
+by refl
+
+lemma adj_iff_exists_edge {v w : V} :
+  G.adj v w ↔ v ≠ w ∧ ∃ (e ∈ G.edge_set), v ∈ e ∧ w ∈ e :=
```
The equivalence `G.adj v w ↔ ∃ (e ∈ G.edge_set), v ∈ e ∧ w ∈ e` also holds, right? That might be a simpler lemma (and together with `ne_of_adj` you still get both components of the RHS).

kmill replied:
I have to periodically remind myself about why this needs the `v ≠ w`: it's because if `v = w`, then `∃ (e ∈ G.edge_set), v ∈ e ∧ w ∈ e` is satisfied by every edge incident to `v`.

fpvandoorn replied:
Ah, you are right! Given that this has confused me (and apparently also you multiple times), do you mind adding that as a docstring to the lemma?

kmill replied:
Done

fpvandoorn reviewed (commented):
This was marked as request review, so here are two more comments.
Are there any other unresolved issues by other reviewers? Otherwise, I'm happy to accept it after these changed, this has (easily) had enough rounds of review by now.

kmill committed:
Merge branch 'master' into graphs_and_bigraphs

kmill committed:
incorporated fpvandoorn suggestion

kmill committed:
added a docstring for a frequently confusing lemma
